test/VaultInventoryReporter.ts

```diff
@@ -0,0 +1,1225 @@
+
+            const deadline = maxDeadline;
+
+            const permitData: InventoryPermitData = {
+                owner: user.address,
+                target: other.address,
+                vault: vault.address,
+                nonce: 0,
+                deadline
+            };
+
+            const sig = await createInventoryPermitSignature(
+                reporter.address,
+                NAME,
+                permitData,
+                user
+            );
+
+            // Make sure user is not approved
+            expect(await reporter.isOwnerOrApproved(vault.address, other.address)).to.be.false;
+
+            await reporter.connect(user).permit(
+                user.address,
+                other.address,
+                vault.address,
+                deadline,
+                sig.v,
+                sig.r,
+                sig.s
+            );
+
+            // Make sure user is approved
+            expect(await reporter.isOwnerOrApproved(vault.address, other.address)).to.be.true;
+
+            // Unapprove user
+            await reporter.connect(user).setApproval(vault.address, ZERO_ADDRESS);
+            expect(await reporter.isOwnerOrApproved(vault.address, other.address)).to.be.false;
+
+            // Try again
+            await expect(
+                reporter.connect(other).permit(
+                    user.address,
+                    other.address,
+                    vault.address,
+                    deadline,
+                    sig.v,
+                    sig.r,
+                    sig.s
+                )
+            ).to.be.revertedWith("VIR_InvalidPermitSignature");
+        });
+
+        it("should not allow a non-admin to set a global approval", async () => {
+            const { reporter, other } = ctx;
+
+            await expect(
+                reporter.connect(other).setGlobalApproval(other.address, true)
+            ).to.be.revertedWith("Ownable");
+        });
+
+        it("should allow a contract owner to set a global approval", async () => {
+            const { reporter, user, other, vault } = ctx;
+
+            expect(await reporter.isGloballyApproved(other.address)).to.be.false;
+
+            await expect(
+                reporter.connect(user).setGlobalApproval(other.address, true)
+            ).to.emit(reporter, "SetGlobalApproval")
+                .withArgs(other.address, true);
+
+            expect(await reporter.isGloballyApproved(other.address)).to.be.true;
+
+            // Make sure can perform reporting
+            await expect(
+                reporter.connect(other).add(vault.address, defaultItems)
+            ).to.not.be.reverted;
+
+            await expect(
+                reporter.connect(user).setGlobalApproval(other.address, false)
+            ).to.emit(reporter, "SetGlobalApproval")
+                .withArgs(other.address, false);
+
+            expect(await reporter.isGloballyApproved(other.address)).to.be.false;
+        });
```

Comment:
A test could be added showing a revoked approval. Say the owner gives approval then later wants to remove for other.address.

Reply:
Good call, added